test/unit/app/controllers/permissions/permissions-controller-test.js

```diff
@@ -0,0 +1,1392 @@
+      )
+    })
+  })
+
+  describe('approvePermissionsRequest', function () {
+
+    let permController, mockRequestUserApproval
+
+    beforeEach(function () {
+      permController = initPermController()
+      mockRequestUserApproval = getMockRequestUserApprovalFunction(
+        permController
+      )
+    })
+
+    it('does nothing if called on non-existing request', async function () {
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty on init',
+      )
+
+      permController.finalizePermissionsRequest = () => {
+        throw new Error('should not be reached')
+      }
+
+      const request = PERMS.approvedRequest(REQUEST_IDS.a, null)
+
+      assert.doesNotReject(
+        permController.approvePermissionsRequest(request, null),
+        'should not throw on non-existing request'
+      )
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals still empty after request',
+      )
+    })
+
+    it('rejects invalid requests', async function () {
+
+      let request
+
+      // bad accounts param
+
+      request = PERMS.approvedRequest(REQUEST_IDS.a, PERMS.requests.eth_accounts())
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.a),
+        ERRORS.validatePermittedAccounts.invalidParam(),
+        'should reject bad accounts'
+      )
+
+      // bad param causing above rejection is here
+      await permController.approvePermissionsRequest(request, null)
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after rejection',
+      )
+
+      // no permissions
+
+      request = PERMS.approvedRequest(REQUEST_IDS.a, {})
+
+      assert.rejects(
+        mockRequestUserApproval(REQUEST_IDS.a),
+        ERRORS.approvePermissionsRequest.noPermsRequested(),
+        'should reject if no permissions in request'
+      )
+
+      await permController.approvePermissionsRequest(request, ACCOUNT_ARRAYS.a)
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after rejection',
+      )
+    })
+
+    it('approves valid request', async function () {
+
+      const request = PERMS.approvedRequest(REQUEST_IDS.a, PERMS.requests.eth_accounts())
+
+      let perms
+
+      assert.doesNotReject(
+        async () => {
+          perms = await mockRequestUserApproval(REQUEST_IDS.a)
+        },
+        'should not reject single valid request'
+      )
+
+      await permController.approvePermissionsRequest(request, ACCOUNT_ARRAYS.a)
+
+      assert.deepEqual(
+        perms, PERMS.finalizedRequests.eth_accounts(ACCOUNT_ARRAYS.a),
+        'produced expected approved permissions'
+      )
+
+      assert.equal(
+        permController.pendingApprovals.size, 0,
+        'pending approvals empty after approval',
+      )
+    })
+
+    it('approves valid requests regardless of order', async function () {
```

Comment:
Hm. For the sake of my own thinking and so that we're both on the same page, the behavior is this:

- Each successful permissions request will return the newly granted permissions _and_ any previously granted permissions
  -  Once the permissions are approved, it basically takes a snapshot of the permissions for the requesting domain, and then returns that
- If a request contains already-granted permissions, the existing permissions with the same method names (`parentCapability` in `rpc-cap` speak) will be overwritten
  - All other already-granted permissions will be unaffected

You're saying that we want a test case with multiple requests, with at least one overwrite, and check the returned permissions for each?

Reply:
Yes, I think that's what I'm saying, but specifically in the case where they're approved out of order. I thought that would be a useful case to test because the behavior is correct but potentially unintuitive, which is the sort of thing that's easy to see breaking in the future (e.g. if things were refactored such that this snapshot of permissions was taken when the request was received rather than when it was approved)

Reply:
As discussed elsewhere, added test case in `permissions-middleware-test` in `wallet_requestPermissions` `describe` block.

Reply:
Ah, they can't be approved out of order, because there can only be a single request pending at any one time 😎

Reply:
Haha oh right! I forgot.